contracts/cw-ics721-bridge/src/ibc.rs

```diff
@@ -0,0 +1,416 @@
+};
+use crate::msg::ExecuteMsg;
+use crate::state::{CLASS_ID_TO_NFT_CONTRACT, ESCROW_CODE_ID, NFT_CONTRACT_TO_CLASS_ID};
+use crate::{state::CHANNELS, ContractError};
+
+#[derive(Serialize, Deserialize, JsonSchema)]
+#[allow(non_snake_case)]
+pub struct NonFungibleTokenPacketData {
+    /// Uniquely identifies the collection which the tokens being
+    /// transfered belong to on the sending chain.
+    pub classId: String,
+    /// URL that points to metadata about the collection. This is not
+    /// validated.
+    pub classUri: Option<String>,
+    /// Uniquely identifies the tokens in the NFT collection being
+    /// transfered.
+    pub tokenIds: Vec<String>,
+    /// URL that points to metadata for each token being
+    /// transfered. `tokenUris[N]` should hold the metadata for
+    /// `tokenIds[N]` and both lists should have the same length.
+    pub tokenUris: Vec<String>,
+    /// The address sending the tokens on the sending chain.
+    pub sender: String,
+    /// The address that should receive the tokens on the receiving
+    /// chain.
+    pub receiver: String,
+}
+
+pub const IBC_VERSION: &str = "ics721-1";
+const ACK_ERROR_FALLBACK: &str =
+    "an unexpected error occurred - error text is hidden because it would serialize as ACK success";
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_open(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelOpenMsg,
+) -> Result<(), ContractError> {
+    validate_order_and_version(msg.channel(), msg.counterparty_version())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_connect(
+    deps: DepsMut,
+    env: Env,
+    msg: IbcChannelConnectMsg,
+) -> Result<IbcBasicResponse, ContractError> {
+    validate_order_and_version(msg.channel(), msg.counterparty_version())?;
+
+    let message = ics_escrow::msg::InstantiateMsg {
+        admin_address: env.contract.address.into_string(),
+        channel_id: msg.channel().endpoint.channel_id.clone(),
+    };
+    let message = WasmMsg::Instantiate {
+        admin: None,
+        code_id: ESCROW_CODE_ID.load(deps.storage)?,
+        msg: to_binary(&message)?,
+        funds: vec![],
+        label: format!(
+            "channel ({}) ICS721 escrow",
+            msg.channel().endpoint.channel_id
+        ),
+    };
+    let message = SubMsg::<Empty>::reply_always(message, INSTANTIATE_ESCROW_REPLY_ID);
+
+    Ok(IbcBasicResponse::new()
+        .add_attribute("method", "ibc_channel_connect")
+        .add_attribute("channel", &msg.channel().endpoint.channel_id)
+        .add_attribute("port", &msg.channel().endpoint.port_id)
+        .add_submessage(message))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_close(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelCloseMsg,
+) -> Result<IbcBasicResponse, ContractError> {
+    match msg {
+        IbcChannelCloseMsg::CloseInit { channel: _ } => Err(ContractError::CantCloseChannel {}),
```

Comment:
We need to de-scrow stuff here
But can do this is another PR

Reply:
Do we? or should we have a governance sudo method that can restore closed channels ?  This will happen more than you think currently for ics20 a proposal is submitted on chain to replace a closed channel for a valid one. 

Reason: Channel trust period very low and failing to update light client states during this period can lead to the channel to be expired.

Reply:
@jhernandezb, governance sudo methods for such things are a GREAT idea... but you're right, this maybe at the light client level. Making an issue to capture this: https://github.com/public-awesome/ics721/issues/25

Reply:
@JakeHartnell @jhernandezb per today's convo seems like when a channel is closed we just leave the NFTs in a channel, don't need to do anything?

Reply:
Resolving this. Think we need some more information about how closes behave and are initiated before we can make concrete design decisions. That information will emerge in the integration testing we can do once this PR lands.